common/logger/src/lib.rs

```diff
@@ -1,74 +1,426 @@
 // Copyright (c) The Libra Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
-#![forbid(unsafe_code)]
-
-use chrono::Utc;
+pub use chrono::Utc;
 use env_logger::filter;
 pub use log::{self, Level, Log, Metadata, Record};
+pub use serde_json;
+
+use serde::Serialize;
+use serde_json::Value;
 use std::{
+    collections::HashMap,
     env,
     fmt::{self, Write},
-    sync::mpsc::{self, Receiver, RecvError, SyncSender, TrySendError},
+    fs::{File, OpenOptions},
+    io,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        mpsc::{self, Receiver, RecvError, SyncSender, TrySendError},
+    },
     thread,
 };
 
 pub mod prelude {
     pub use crate::{crit, debug, error, info, trace, warn};
 }
+pub use std::io::Write as IoWrite;
 
 pub const CHANNEL_SIZE: usize = 256;
 pub const DEFAULT_TARGET: &str = "libra";
 const RUST_LOG: &str = "RUST_LOG";
 
+pub trait StructLogSink: Sync {
+    fn send(&self, entry: StructuredLogEntry);
+}
+
+// This is poor's man AtomicReference from crossbeam
+// It have few unsafe lines, but does not require extra dependency
+static NOP: NopStructLog = NopStructLog {};
+static mut STRUCT_LOGGER: &'static dyn StructLogSink = &NOP;
+static STRUCT_LOGGER_STATE: AtomicUsize = AtomicUsize::new(UNINITIALIZED);
+const UNINITIALIZED: usize = 0;
+
+const INITIALIZING: usize = 1;
+const INITIALIZED: usize = 2;
+
+#[derive(Default, Serialize)]
+pub struct StructuredLogEntry {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    text: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pattern: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    name: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    module: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    location: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    git_rev: Option<&'static str>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    timestamp: Option<String>,
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    data: HashMap<&'static str, Value>,
+}
+
+#[must_use = "use StructuredLogEntry::send to send structured log"]
+impl StructuredLogEntry {
```

Comment:
have you considered making the fields above public? Does it make sense for this struct to be public?

Reply:
This struct has to be public since it is initialized with macro